br/pkg/lightning/backend/external/kv_reader.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+
+	"github.com/pingcap/tidb/br/pkg/storage"
+)
+
+type kvReader struct {
+	byteReader *byteReader
+}
+
+func newKVReader(ctx context.Context, name string, store storage.ExternalStorage, initFileOffset uint64, bufSize int) (*kvReader, error) {
+	sr, err := openStoreReaderAndSeek(ctx, store, name, initFileOffset)
+	if err != nil {
+		return nil, err
+	}
+	br, err := newByteReader(ctx, sr, bufSize)
+	return &kvReader{
+		byteReader: br,
+	}, nil
+}
+
+func (r *kvReader) nextKV() (key, val []byte, err error) {
```

Comment:
kvReader and statReader's only have different nextxxx function, can we merge the code into one? just pass a function into the next function.

```
var k, v
next(func() {
 k, v = 
})

Reply:
Merging them into one is a bad idea because one may mistakenly call `nextProps()` after `nextKV()`, which is nonsense.